search-contests-#83-and-dashboard-visibility-#84

yliu298 opened the pull request:
### What this PR does (required):
- set up dashboard can be visited by any user who logged in or not
- set up search bar with simple search by title, and advance search by title, start date, and end date

### Screenshots / Videos (required):
- dashboard page without logged in
![Screenshot 2021-07-30 183925](https://user-images.githubusercontent.com/65554508/127718696-309618dc-921e-4ad2-acdf-f96a7121e841.jpg)
- Simple search
![Screenshot 2021-07-30 183925](https://user-images.githubusercontent.com/65554508/127718753-f5c91f1a-6b27-4a76-9886-a14e3a19e68c.jpg)
- Advantage search
![Screenshot 2021-07-30 184052](https://user-images.githubusercontent.com/65554508/127718769-07f9592e-42f0-4f22-a73b-65907390d558.jpg)
![Screenshot 2021-07-30 184120](https://user-images.githubusercontent.com/65554508/127718784-6ca24491-2179-472d-86b7-431445d9fccd.jpg)


### Any information needed to test this feature (required):
- simple search-> type any string and click search button
-  advance search-> click advance search button-> type any string, select a date range between start date and end date

yliu298 committed:
search-contests-#83-and-dashboard-visibility-#84

yliu298 committed:
Merge branch 'main' into search-contests-#83

yliu298 committed:
fixed conflict files

rajivtitus commented on client/src/pages/Discovery/Discovery.tsx:
```diff
@@ -31,11 +43,75 @@ export default function Discovery(): JSX.Element {
+    }>,
+  ) => {
+    getAllContestsBySimpleSearch(search).then((data) => {
+      if (data.contest.length) {
```
Send the response as `success` | `error` from the BE would make more sense :)

Then it would be `if (data.success)`

yliu298 replied:
modified contest controller

rajivtitus commented on client/src/pages/Discovery/Discovery.tsx:
```diff
@@ -31,11 +43,75 @@ export default function Discovery(): JSX.Element {
+    }>,
+  ) => {
+    getAllContestsByAdvanceSearch(title, startTime, endTime).then((data) => {
+      if (data.contest.length) {
```
Please see above

yliu298 replied:
modified contest controller

rajivtitus commented on client/src/pages/Discovery/SearchForm/SimpleSearch/SimpleSearch.tsx:
```diff
@@ -0,0 +1,63 @@
+    >
+      {({ handleSubmit, handleChange, values, touched, errors, isSubmitting }) => (
+        <form onSubmit={handleSubmit} noValidate>
+          <div className={classes.root}>
```
The MUI Box component already has flexbox built into it. Please consider using that instead of using classes to style divs.

yliu298 replied:
changed to use Box components

rajivtitus commented on client/src/pages/Discovery/SearchForm/SimpleSearch/SimpleSearch.tsx:
```diff
@@ -0,0 +1,63 @@
+      {({ handleSubmit, handleChange, values, touched, errors, isSubmitting }) => (
+        <form onSubmit={handleSubmit} noValidate>
+          <div className={classes.root}>
+            <div className={classes.search}>
```
Maybe use MUI Box components to stay consistent with code base?

Ditto throughout

yliu298 replied:
changed to use Box components

rajivtitus commented on client/src/pages/Discovery/SearchForm/SimpleSearch/SimpleSearch.tsx:
```diff
@@ -0,0 +1,63 @@
+        <form onSubmit={handleSubmit} noValidate>
+          <div className={classes.root}>
+            <div className={classes.search}>
+              <div className={classes.searchIcon}>
```
Same as comment above for flexbox using Box component

yliu298 replied:
changed to use Box components

rajivtitus commented on server/routes/contest.js:
```diff
@@ -8,13 +8,17 @@ const {
-router.get('/all/contests', protect, getContests);
+router.get('/all/contests', getContests);
 router.post('/:id/submission', protect, createSubmissionByContestId);
+router.post('/all/contests/simpleSearch', getContestsBySimpleSearch);
```
The current solution is functional, good work with that. However, we have two new endpoints just for searching, maybe consider the usage of query parameters in the route (That way we don't need to create any new routes) :)

yliu298 replied:
changed to using one new endpoint, and add type null as parameter to determine which search should be used in controller

rajivtitus commented on server/controllers/submission.js:
```diff
@@ -19,25 +19,132 @@ exports.getAllSubmissions = asyncHandler(async (req, res, next) => {
+        res.status(200).json({ submission: submissionList, isOwner: false });
+      }
+    } else {
+      const submission = await Submission.findOne({ userId: userId }).populate('userId');
```
Isn't this the same as the if/else clause beginning on Line 61?

yliu298 replied:
there are in difference if/else block. the reason is to check the situations: 1, the user is not the contest owner and has no submissions. 2, he is not the owner but he has submissions. 3, he is the owner but no submission. 4, he is the owner and he has submissions

rajivtitus replied:
As we discussed on our call this morning, if a user is a contest creator - they should not have the ability to submit designs to their own contest (so you will need to change this accordingly)

yliu298 committed:
updated base on suggestion

rajivtitus reviewed (approved):
Thanks for making some of the requested changes. 

As for the route, we spoke about the usage of route parameters which has not been implemented in the changes. However, since the current solution is also functional, I am approving this PR. (For future PRs, please make use of route parameters rather than having multiple api endpoints)

yliu298 committed:
Merge branch 'main' into search-contests-#83

yliu298 committed:
updated query string and controller

yliu298 commented:
> Thanks for making some of the requested changes.
> 
> As for the route, we spoke about the usage of route parameters which has not been implemented in the changes. However, since the current solution is also functional, I am approving this PR. (For future PRs, please make use of route parameters rather than having multiple api endpoints)

Updated to use query string in one endpoint which getAllcontests, removed all new endpoints, and set visibility on submit design button.

rajivtitus commented on client/src/pages/Discovery/Discovery.tsx:
```diff
@@ -30,14 +30,18 @@ export default function Discovery(): JSX.Element {
-    });
-  }, [updateSnackBarMessage]);
+    const fetchAllcontests = () => {
+      history.push(`/discovery`);
```
Why are we pushing the '/discovery' route onto the history stack as this useEffect will only run when the user is on the discovery page?

rajivtitus commented on server/controllers/contest.js:
```diff
@@ -63,20 +63,67 @@ exports.getContest = asyncHandler(async (req, res, next) => {
+        contestList.push(contestData);
+      });
+      res.status(200).json({ contest: contestList });
+    } else if (search !== 'null') {
```
Wouldn't we also want to check if the startTime/endTime are null over here?

yliu298 replied:
I think that it doesn't need to check because there are difference handle submit function: handleSimpleSubmit and handleSubmit, in handleSimpleSubmit , only the search params won't be null, and in handleSubmit, search params must be null.

rajivtitus commented on client/src/pages/Discovery/Discovery.tsx:
```diff
@@ -51,7 +55,8 @@ export default function Discovery(): JSX.Element {
     }>,
   ) => {
-    getAllContestsByAdvanceSearch(search, null, null, null).then((data) => {
+    history.push(`/discovery?search=${search}`);
```
Why are we pushing these routes onto the history stack here and below? :)

yliu298 replied:
removed discovery, was thinking the path is  in the discovery page

yliu298 committed:
updated query string path

rajivtitus reviewed (commented):
Changes look good to me, thanks!
